### PR TITLE
Memory commit threshold in MB at which to create a dump with number of dumps need taken.

### PR DESCRIPTION
**Pull request by kishanAnem**
but problem with CommandLineOptions. it isn't letting a  set -m. because  we already gave 'm' for dump type. it is comparing char level i think. pls suggest some options. just kept z and y.

### CODE REVIEW

**Comment by lowleveldesign** [comment]
@kishan3034, I looked through the code and it looks good. I need to run it through some tests and I will do that over this week.

Regarding the command line options. I see the problem with -m and you are right that it might be too hard to implement. So maybe for cases when we break the compatibility with procdump we should use the long options ("--minmem" and "--maxmem"). It is a simple change and we keep letters y and z free, in case they appear in procudmp. Sounds OK?

**Comment by kishanAnem** [comment]
Thanks @lowleveldesign.

File: MiniDumper/MiniDumper.cs [hunk 1]
```diff
@@ -178,14 +200,85 @@ public void PrintDebugString(OUTPUT_DEBUG_STRING_INFO outputDbgStrInfo)
...
+            OnTimedEvent();
+        }
+
+        private void OnTimedEvent()
+        {
+            processCommit = GetProcessCommit(hProcess);
+            if ((CommitThreshold.HasValue && processCommit >= CommitThreshold) || (CommitDrops.HasValue && processCommit >= CommitDrops))
+            {
+                Debug.Write(CommitThreshold.HasValue ? $"Commit:\t >= {CommitDrops / 1024 / 1024}MB" : $"Commit:\t <= {CommitDrops / 1024 / 1024}MB");
+                Console.WriteLine("Commit {0}", processCommit / 1024 / 1024);
```
Comment (lowleveldesign) [inline MiniDumper/MiniDumper.cs]: Please make this message more meaningful (as in the line above)

File: MiniDumper/Debugger.cs [hunk 2]
```diff
@@ -48,41 +53,72 @@ void TakeDumps()
...
+                if (_options.MemoryCommitThreshold.HasValue || _options.MemoryCommitDrops.HasValue)
+                {
+                    bool isTimerSet = false;
+                    if (!_options.NeedAttachDebugger)
+                    {
+                        DetachProcess();
+
+                        while (!IsPollingCompleted)
+                        {
+                            if (!isTimerSet)
```
Comment (lowleveldesign) [inline MiniDumper/Debugger.cs:86]: Even for single-line conditions please use parenthesis - they are easier to spot and that's the convention we use in minidumper code. There are few pleaces in the code where it needs to be fixed

File: MiniDumper/MiniDumper.cs [hunk 3]
```diff
@@ -178,14 +200,85 @@ public void PrintDebugString(OUTPUT_DEBUG_STRING_INFO outputDbgStrInfo)
...
 
+        public void MemoryCommitThreshold(uint? commitThreshold, uint? commitDrops, int NumberOfDumps, ref bool isTimerSet)
+        {
+            isTimerSet = true;
+            CommitThreshold = commitThreshold * 1048576;
+            CommitDrops = commitDrops * 1048576;
+
+            hProcess = ProcessNativeMethod.OpenProcess(ProcessAccessFlags.QueryInformation, false, pid);
+            if (hProcess.IsInvalid)
+                throw new ArgumentException(String.Format("Unable to open process {0}, error {x:8}", pid, Marshal.GetLastWin32Error()));
```
Comment (lowleveldesign) [inline MiniDumper/MiniDumper.cs]: The "{x:8}" format string is wrong - I guess you meant "{1:x8}". There are two more places with the same issue - please check VS warnings.

File: MiniDumper/MiniDumper.cs [hunk 4]
```diff
@@ -217,9 +310,15 @@ public int NumberOfDumpsTaken
...
         }
 
+        public bool IsProcessTerminated { get => isProcessTerminated; }
+
+
+
         public void Dispose()
         {
             target.Dispose();
+            hProcess.Dispose();
```
Comment (lowleveldesign) [inline MiniDumper/MiniDumper.cs]: Please check if hProcess is not null - we don't use it in other situations than memory commits checks and it throws NullPointerException.

File: MiniDumper/CommandLineOptions.cs [hunk 5]
```diff
@@ -28,6 +28,12 @@ class CommandLineOptions
         [Option('l', HelpText = "Display the debug logging of the process + diagnostics info from the minidumper.")]
         public bool Verbose { get; set; }
 
+        [Option('y', HelpText = "Memory commit threshold in MB at which to create a dump.")]
```
Comment (lowleveldesign) [inline MiniDumper/CommandLineOptions.cs]: Let's change this to "maxmem" instead of 'y'.

File: MiniDumper/CommandLineOptions.cs [hunk 6]
```diff
@@ -28,6 +28,12 @@ class CommandLineOptions
...
+        public uint? MemoryCommitThreshold { get; set; }
+        
+        [Option('z', HelpText = "Memory commit threshold in MB at which to create a dump.")]
```
Comment (lowleveldesign) [inline MiniDumper/CommandLineOptions.cs]: Please change 'z' to "minmem" and also the help text is wrong, should be: "Trigger when memory commit drops below specified MB value."

File: MiniDumper/MiniDumper.cs [hunk 7]
```diff
@@ -178,14 +200,85 @@ public void PrintDebugString(OUTPUT_DEBUG_STRING_INFO outputDbgStrInfo)
             }
         }
 
+        public void MemoryCommitThreshold(uint? commitThreshold, uint? commitDrops, int NumberOfDumps, ref bool isTimerSet)
```
Comment (lowleveldesign) [inline MiniDumper/MiniDumper.cs]: For convention please change this method name to "DumpOnMemoryCommitThreshold".

**Review by lowleveldesign — CHANGES_REQUESTED** [review]
I was a little bit too fast with my previous comment. I run few tests and also had a look again at the code. Please check the comments. Thanks

File: MiniDumper/MiniDumper.cs (hunk 1 above)
Comment (kishanAnem) [reply MiniDumper/MiniDumper.cs]: I just followed procdump. please suggest some message .
image

Comment (lowleveldesign) [reply MiniDumper/MiniDumper.cs]: Ach, OK. That is fine then, but please add 'Mb' at the end. Someone might not know what is the memory unit.

**Review by lowleveldesign — APPROVED** [review]

**Comment by lowleveldesign** [comment]
@kishan3034 thanks, I will merge the changes. There is still a problem with the format strings (please check warnings reported by VS), but I will fix them. It should be {1:x8}. I need to make small refactoring of the main class and will make a release soon.

**Comment by kishanAnem** [comment]
@lowleveldesign  a big thanks. I will try to fix those warnings.

**Comment by lowleveldesign** [comment]
@kishan3034 It's all right - I already fixed them.

**Comment by kishanAnem** [comment]
@lowleveldesign big thanks it isn't showing  now.

**Pull request merged**